### DISCUSSION TITLE
Inquiry about energest and low-power modes in Contiki-NG Cooja

### DISCUSSION THREAD

**Discussion by aceylan** [contiki-ng/contiki-ng#2820] [Q&A]
Hello. 

 I am currently working with Contiki-NG and experimenting with energest and simple-energest in Cooja for power consumption analysis. However, I have encountered an issue where the Low Power Mode (LPM) always shows 0 when using these tools in Cooja.

I would like to inquire if there is a specific procedure or code snippet to enable LPM mode in Cooja for accurate power consumption simulation. Are there any settings or configurations that I might be missing?

Additionally, I am curious whether testing in a real-world environment is necessary to observe LPM transitions accurately, or if there are ways to simulate low-power scenarios within the Cooja environment.

Any guidance, suggestions, or insights on this matter would be greatly appreciated.

Thank you in advance for your time and assistance.

**Comment by aceylan** [comment]
does anybody know it?
How can I measure lpm on tsch? I made it on csma mode with 
NETSTACK_MAC.off();
ENERGEST_SWITCH(ENERGEST_TYPE_CPU, ENERGEST_TYPE_LPM);
codes. is it the right way?

I had assumed that the Low Power Mode (LPM) would be activated automatically. Do I need to use the following command when measuring energy on TSCH as well?
`ENERGEST_SWITCH(ENERGEST_TYPE_CPU, ENERGEST_TYPE_LPM);
`Can you clarify if this command is necessary for TSCH or if there are alternative approaches for enabling LPM mode during energy measurements in this context?

**Comment by chakibouzane** [comment]
> I am currently working with Contiki-NG and experimenting with energest and simple-energest in Cooja for power consumption analysis. However, I have encountered an issue where the Low Power Mode (LPM) always shows 0 when using these tools in Cooja.

Hello
I ran into the same issue, and all my energest values were always 0 in Cooja. after a lot of digging, I figured out a couple of things that might help you/whoever is reading this:

- Don’t use Cooja motes for energy measurements  
  Cooja motes don’t emulate real hardware, so energest will always log 0 for everything.  
  Use **Z1** or **Sky** motes instead, these emulate hardware and return energy consumption values.

- Energest is not enabled by default  
  It took me hours to figure this out. You have to explicitly enable it, otherwise all values will be 0.
  There are two ways to enable it:
  1. Globally: edit `os/sys/energest.h` and set  
     ```c
     #define ENERGEST_CONF_ON 1
     ```
     This will enable energest for all simulations

  2. Per project:  by adding
     ```c
     #define ENERGEST_CONF_ON 1
     ```
     to the `project-conf.h` file in the project directory, create it if it doesn't exist.  
     I did a custom simulation and I used this method: [https://github.com/ch33kaboo/RPL-DIS-attack](https://github.com/ch33kaboo/RPL-DIS-attack)  
     You can check out the `project-conf.h` files there.

I wish this was mentioned more clearly in the docs, I spent way too long trying to figure out why everything was zero. I even had to go through the source code in `os/sys/energest.c` and `os/sys/energest.h` before I found this bit:
```
#ifndef ENERGEST_CONF_ON
/* Energest is disabled by default */
#define ENERGEST_CONF_ON 0
#endif /* ENERGEST_CONF_ON */
```
So, hopefully this saves the next person some time :)
